cmd/kubeadm/app/apis/bootstraptoken/v1/utils.go

```diff
@@ -30,9 +31,61 @@ import (
 	bootstrapsecretutil "k8s.io/cluster-bootstrap/util/secrets"
 )
 
-// ToSecret converts the given BootstrapToken object to its Secret representation that
+// MarshalJSON implements the json.Marshaler interface.
+func (bts BootstrapTokenString) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf(`"%s"`, bts.String())), nil
+}
+
+// UnmarshalJSON implements the json.Unmarshaller interface.
+func (bts *BootstrapTokenString) UnmarshalJSON(b []byte) error {
+	// If the token is represented as "", just return quickly without an error
+	if len(b) == 0 {
+		return nil
+	}
+
+	// Remove unnecessary " characters coming from the JSON parser
+	token := strings.Replace(string(b), `"`, ``, -1)
+	// Convert the string Token to a BootstrapTokenString object
+	newbts, err := NewBootstrapTokenString(token)
+	if err != nil {
+		return err
+	}
+	bts.ID = newbts.ID
+	bts.Secret = newbts.Secret
+	return nil
+}
+
+// String returns the string representation of the BootstrapTokenString
+func (bts BootstrapTokenString) String() string {
+	if len(bts.ID) > 0 && len(bts.Secret) > 0 {
+		return bootstraputil.TokenFromIDAndSecret(bts.ID, bts.Secret)
+	}
+	return ""
+}
+
+// NewBootstrapTokenString converts the given Bootstrap Token as a string
+// to the BootstrapTokenString object used for serialization/deserialization
+// and internal usage. It also automatically validates that the given token
+// is of the right format
+func NewBootstrapTokenString(token string) (*BootstrapTokenString, error) {
+	substrs := bootstraputil.BootstrapTokenRegexp.FindStringSubmatch(token)
```

Comment:
Overall LGTM
/lgtm

When I go thought the code, I find `IsValidBootstrapToken` , `BootstrapTokenRegexp` are only used in `/cmd/`

Not sure if we need to move Regexp and isVaild func above to this utils.

And there is an old comment:
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/cluster-bootstrap/token/util/helpers.go#L30

Reply:
@dixudx any ideas?

Reply:
originally we wanted the kubeadm bootstraptoken utils to live in k8s.io/cluster-bootstrap, but the idea was rejected with the argument that these introduce wrapper APIs that are kubeadm specific. only the bootstrap token Secret format matters as a specification and not the kubeadm originated BootstrapTokenString.

so what we ended up having is:
- generic utils that can be used outside of kubeadm and residing in a library under k8s.io/cluster-bootstrap
- kubeadm specific code in /cmd/kubeadm/...api including the one responsible converting token data <-> secret

overall what is in k8s.io/cluster-bootstrap at this point is a library API at this point and moving it can be a breaking change to users, so we rather not do it.

things like BootstrapTokenRegexp and IsValidBootstrapToken makes sense to be in k8s.io/cluster-bootstrap. if someone wishes to use bootstrap tokens with their own bootstrap mechanism and without kubeadm.
the kubeadm token data <-> secret utils are quite useful but were not accepted in k8s.io/cluster-bootstrap.

Reply:
> // TODO(dixudx): refactor this to util/secrets and util/tokens 

this TODO seems still relevant as something that can be done in `/staging/... /cluster-bootstrap`,  e.g. having to separate files `tokens.go` and `secrets.go`. i don't remember the discussions about it.